### PR TITLE
[GLES] Improve subtitle rendering

### PR DESCRIPTION
**Pull request by kszaq**
## Description
Most GLES hardware support BGRA textures via an extension. If supported, use it to avoid doing malloc for convering textures.

## Motivation and Context
With some kind of subtitles there is a visible frame skip when subtitles are rendered. This is especially visible when subtitle is a 1920x1080 picture:
https://forum.kodi.tv/showthread.php?tid=315216
https://forum.libreelec.tv/thread/8100-bug-s905x-krypton-frame-skips-caused-by-burned-in-forced-subtitles/

## How Has This Been Tested?
LibreELEC on Amlogic S905 (GLES 2.0 with GL_EXT_texture_format_BGRA8888) and S912 (GLES 3.0 and GL_EXT_texture_format_BGRA8888)

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by wrxtasy** [comment]
Nice work @kszaq 
Tested with Linux (CE) Kodi Leia - MINIX U9: (GLES 3.0)

Yes this make a noticeable difference when Overlays like Subtitles are first shown onscreen.
The video frame skip is no longer there after this PR.

**Comment by lrusak** [comment]
How is this any different than #15505 ?? It wasn't accepted there, why do you think it would be accepted again? :confused:

**Comment by kszaq** [comment]
#15505 was a workaround for broken headers, this PR introduces suport for unpack and BGRA extensions to OverlayRenderGL. I think I described it quite thoroughly. 

Your comments from #15505 have been addressed. I am aware that my code style may not be the best, I will happily accept all suggestions.

**Comment by lrusak** [comment]
> #15505 was a workaround for broken headers, this PR introduces suport for unpack and BGRA extensions to OverlayRenderGL. I think I described it quite thoroughly.
> 
> Your comments from #15505 have been addressed.

You are still working around broken headers by defining GL defines. Also the extension `GL_EXT_unpack_subimage` only adds support for the `_EXT` versions like `GL_UNPACK_ROW_LENGTH_EXT` and not `GL_UNPACK_ROW_LENGTH`. Even though they both have the same definition one is defined in `GLES2/gl2ext.h` and the other in `GLES3/gl3.h` and I'd rather be explicit about it.

Overall I'd really rather wait for any more GLES3 changes until #13457 is merged and we can split GLES2 and GLES3 apart.

**Comment by kszaq** [comment]
@lrusak Understood. I have now adjusted the PR to only add support for BGRA textures in OverlayRendererGL for GLES. This also solves the frame drop issue.

**Comment by popcornmix** [comment]
@kszaq do you have a link to a sample file that shows the problem? (linked threads have dead download links).
I'd like to test it on a Pi.

**Comment by kszaq** [comment]
@popcornmix https://kszaq.libreelec.tv/samples/avatar%20remux%20sample-trim.mkv

**Comment by popcornmix** [comment]
``GL_EXT_texture_format_BGRA8888`` is supported on Pi.

The shortcut this PR provides by using the extension saves ~36ms per new subtitle of this style on a Pi3 which is likely to avoid a frame drop when rendering a new sub.

So, +1 from me.

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererGL.cpp
Comment (lrusak) [inline xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererGL.cpp]: This seems to change the behaviour. Before it we compared stride to width, now you are comparing stride to stride, why?

Comment (kszaq) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererGL.cpp]: `int bytesPerLine = bytesPerPixel * width`, so this is still comparing stride to width. If you look at previous implementation, it wasn't entirely correct because it compared stride (in BYTES) to width (in PIXELS) - but it was okay since this comparison was only made if `alpha == true`, i.e. one byte per pixel. Now we need to take bytes per pixel into consideration. 

In the Avatar sample stride = 7680, width = 1920, bpp = 4, so we really need to compare stride to bytesPerLine.

**Comment by smp79** [comment]
@popcornmix Another [sample](https://www.dropbox.com/s/aypt4ejky3sfv0y/bd_sample.m2ts?dl=1) that is causing frame drops on Pi 3 when subtitles are enabled. I see no improvement with this PR.

**Comment by popcornmix** [comment]
It looks like this PR saves ~63ms per new subtitle which would be about 1.5 frames skipped.
But there is still other work being done in the render thread which means we are still skipping some frames.
@kszaq does this sample play okay for you?

**Comment by kszaq** [comment]
@popcornmix Thank you for testing and sharing your results!

@smp79 With your sample a single frame is dropped every time a new subtitle is rendered. Without this PR 1-3 frames are dropped every subtitle rendered.

**Comment by popcornmix** [comment]
@kszaq have you been looking at this code closely enough to say where the time is spent that causes the remaining frame drop? I'm guessing it's freetype, but I haven't checked for sure.

Just wondering how long it takes and if there's anything more that can be done.

**Comment by kszaq** [comment]
@popcornmix No, I haven't. Perhaps in a few days...

**Comment by kszaq** [comment]
My *guess* is that rendering subtitles unnecessarily block rendering thread. While this PR speeds up rendering, I suspect it would be best to not let overlay renderer block video rendering, late subtitle is not an issue while a dropped frame is. Unfortunately I have no idea how to achieve this.

**Comment by popcornmix** [comment]
Yes anything that takes significant time (a reasonable fraction of a frame time) shouldn't run from render thread but a separate thread. I suspect the code that generates bitmaps from freetype fonts is called from render thread but I haven't investigated.

The test is, if commenting out that call removes the remaining frame skips.

If it does then launching it as a background job and only rendering it when complete (e.g. a few frames later) would be a preferable solution.

**Comment by kszaq** [comment]
In both samples subtitles are in PGS format, so bitmaps and it looks like bitmap conversion and rendering blocks video rendering thread for too long. When I moved `convert_rgba` from OverlayRenderer to DVDOverlayCodecFFmpeg with sample from @smp79 I had no frames dropped but this is certainly not a proper solution. Unfortunately my C++ skills end on modifying single files and I have no idea how to implement threading for rendering subtitles.

**Comment by afl1** [comment]
I came to the same result as @popcormix when I did subtitles issue investigations. Rendering subtitles need own thread and only when rendering is completed can be overlayed. Currently subtitles rendering is done in main thread eating time from frame playback. I think this is principial bug and have to fix by kodi developers. Any other partial fix is only patching this issue.

**Comment by afl1** [comment]
I think there is not only issue with subtitles. Also OSD have to be rendered in in extra thread or subtitles  and OSD can be done asynchronously to frame rendering in extra thread.

**Comment by pkerling** [comment]
> I came to the same result as @popcornmix when I did subtitles issue investigations. Rendering subtitles need own thread and only when rendering is completed can be overlayed.

As said in the thread, this wouldn't change anything for bitmap subs which already seem to be problematic.

> Any other partial fix is only patching this issue.

See above - this is orthogonal. There is no reason to shuffle data around needlessly no matter where and how it is rendered.

**Comment by FernetMenta** [comment]
> Rendering subtitles need own thread and only when rendering is completed can be overlayed

Rendering can only be done by the render thread. That's how GL/GLES work. (maybe you have your own definition of rendering). Texture loading can be offloaded to another thread, if this thread shares the GL context with the render thread.

**Comment by pkerling** [comment]
> (maybe you have your own definition of rendering)

I think it was concerning font rendering (FreeType)

**Comment by kszaq** [comment]
@lrusak Is there any other change I should make to this PR?

**Review by lrusak — APPROVED** [review]
I'm fine with this band-aid if @MartijnKaijser wants to pull this in for the next release. Otherwise it can go into V19

**Comment by kszaq** [comment]
Thank you.

I don't consider this a band-aid as even if the root cause of frame drop is solved, it's always nice to save ca. 50 ms of CPU time on unnecessary conversion.

**Comment by popcornmix** [comment]
I'm in favour of getting this in. In some situations it will avoid the frame skip when a new subtitle appears. In others it should reduce it. More can obviously be done but this is a nice improvement.

**Pull request merged**
